File: core/algorithms/learning_modes.py
"""Unified Phenitime learning-mode taxonomy (mirrors backend/src/utils/learningModes.js)."""

LEARNING_MODES = {
    "THEORY": frozenset({"LT"}),
    "PRACTICE": frozenset({"TH", "PM", "TN", "SB", "XT"}),
    "ONLINE": frozenset({"ONLINE", "ELN", "ONLINE_ELEARNING", "ONLINE_COURSERA"}),
    "SPECIAL": frozenset({"DA", "ĐA", "KL", "TT", "DN", "BV"}),
}

ASYNC_ONLINE_CLASS_TYPES = frozenset({"ONLINE_ELEARNING", "ONLINE_COURSERA"})


def normalize_learning_type(value):
    return str(value or "").strip().upper()


def get_learning_mode(value):
    normalized = normalize_learning_type(value)

    if normalized in LEARNING_MODES["THEORY"]:
        return "THEORY"
    if normalized in LEARNING_MODES["PRACTICE"]:
        return "PRACTICE"
    if normalized in LEARNING_MODES["ONLINE"]:
        return "ONLINE"
    if normalized in LEARNING_MODES["SPECIAL"]:
        return "SPECIAL"

    return "THEORY"


def is_online_room_requirement(room_type_req):
    return normalize_learning_type(room_type_req) == "ONLINE"


def should_skip_scheduling(row):
    room_type_req = normalize_learning_type(
        row.get("room_type_req") or row.get("course_default_room_type") or "LT"
    )
    class_type = normalize_learning_type(row.get("class_type"))

    if room_type_req == "ONLINE" and class_type in ASYNC_ONLINE_CLASS_TYPES:
        return True

    return False

File: backend/src/utils/learningModes.js
/**
 * Unified Phenitime learning-mode taxonomy.
 * Keeps Node generator, Python scheduler, and React UI in sync.
 */

const {
    DELIVERY_CHANNELS,
    resolveDeliveryChannel,
    requiresSchedulingForCourse,
    resolveOnlineSectionClassType,
    isOnlineSectionGroupCode,
    isCourseraPracticeGroupCode,
} = require('./deliveryChannels');

const LEARNING_MODES = {
    THEORY: ['LT'],
    PRACTICE: ['TH', 'PM', 'TN', 'SB', 'XT'],
    ONLINE: ['ONLINE', 'ELN', 'ELN0', 'ONLINE_ELEARNING', 'ONLINE_COURSERA', 'ELEARNING', 'COURSERA'],
    SPECIAL: ['DA', 'ĐA', 'KL', 'TT', 'DN', 'BV', 'SPECIAL'],
};

const ASYNC_ONLINE_CLASS_TYPES = new Set([
    'ELN',
    'ELN0',
    'ONLINE',
    'ONLINE_ELEARNING',
    'ONLINE_COURSERA',
    'ELEARNING',
    'COURSERA',
]);

const SPECIAL_COURSE_NAME_KEYWORDS = [
    'đồ án',
    'do an',
    'thực tập',
    'thuc tap',
    'khóa luận',
    'khoa luan',
];

function normalizeLearningType(type) {
    return String(type ?? '').trim().toUpperCase();
}

function getLearningMode(type) {
    const normalized = normalizeLearningType(type);

    if (LEARNING_MODES.THEORY.includes(normalized)) return 'THEORY';
    if (LEARNING_MODES.PRACTICE.includes(normalized)) return 'PRACTICE';
    if (LEARNING_MODES.ONLINE.includes(normalized)) return 'ONLINE';
    if (LEARNING_MODES.SPECIAL.includes(normalized)) return 'SPECIAL';

    return 'THEORY';
}

function isPracticeRoomType(roomType) {
    return getLearningMode(roomType) === 'PRACTICE';
}

function isOnlineLearningType(type) {
    return getLearningMode(type) === 'ONLINE';
}

function isSpecialLearningType(type) {
    return getLearningMode(type) === 'SPECIAL';
}

function resolveCourseLearningMode(course, defaultRoomType = 'LT') {
    const channel = resolveDeliveryChannel(course);
    if (channel === DELIVERY_CHANNELS.SPECIAL) {
        return 'SPECIAL';
    }
    if (channel === DELIVERY_CHANNELS.ELEARNING || channel === DELIVERY_CHANNELS.COURSERA) {
        return 'ONLINE';
    }

    const roomType = normalizeLearningType(
        course?.default_room_type || course?.room_type || defaultRoomType,
    );

    if (isOnlineLearningType(roomType)) {
        return 'ONLINE';
    }

    if (isSpecialLearningType(roomType)) {
        return 'SPECIAL';
    }

    return getLearningMode(roomType);
}

function resolveOnlineClassType(course) {
    const channel = resolveDeliveryChannel(course);
    return resolveOnlineSectionClassType(channel);
}

function requiresSchedulingForSection(section) {
    if (!section) return false;

    const roomTypeReq = normalizeLearningType(section.room_type_req);
    if (roomTypeReq === 'ONLINE') {
        return false;
    }

    const groupCode = String(section.section_id || '').match(/\(([^)]+)\)$/)?.[1] || '';
    if (isOnlineSectionGroupCode(groupCode) && !isCourseraPracticeGroupCode(groupCode)) {
        return false;
    }

    return requiresSchedulingForCourse(section.course || {});
}

function resolvePracticeClassType(roomTypeReq) {
    const normalized = normalizeLearningType(roomTypeReq);
    if (normalized === 'PM') return 'PM';
    if (normalized === 'TH') return 'TH';
    return 'TH';
}

module.exports = {
    DELIVERY_CHANNELS,
    LEARNING_MODES,
    ASYNC_ONLINE_CLASS_TYPES,
    SPECIAL_COURSE_NAME_KEYWORDS,
    normalizeLearningType,
    getLearningMode,
    isPracticeRoomType,
    isOnlineLearningType,
    isSpecialLearningType,
    resolveCourseLearningMode,
    resolveDeliveryChannel,
    requiresSchedulingForCourse,
    requiresSchedulingForSection,
    resolveOnlineClassType,
    resolvePracticeClassType,
    isOnlineSectionGroupCode,
    isCourseraPracticeGroupCode,
};
